packages/dart_frog_cli/e2e/test/new_test.dart

```diff
@@ -0,0 +1,484 @@
+        exists,
+      );
+    });
+
+    test('Creates middleware in dynamic route', () async {
+      await dartFrogNewMiddleware('/[id]', directory: projectDirectory);
+
+      expect(fileAt('[id]/_middleware.dart', on: routesDirectory), exists);
+    });
+
+    test('Creates middleware in nested dynamic route', () async {
+      await dartFrogNewMiddleware(
+        '/inn/[id]/route',
+        directory: projectDirectory,
+      );
+
+      expect(
+        fileAt('inn/[id]/route/_middleware.dart', on: routesDirectory),
+        exists,
+      );
+    });
+
+    test('Creates middleware in existing file route', () async {
+      await dartFrogNewRoute(
+        '/existing_file_route',
+        directory: projectDirectory,
+      );
+
+      expect(fileAt('existing_file_route.dart', on: routesDirectory), exists);
+
+      await dartFrogNewMiddleware(
+        '/existing_file_route',
+        directory: projectDirectory,
+      );
+
+      expect(
+        fileAt('existing_file_route.dart', on: routesDirectory),
+        doesNotExist,
+      );
+
+      expect(
+        fileAt('existing_file_route/index.dart', on: routesDirectory),
+        exists,
+      );
+
+      expect(
+        fileAt('existing_file_route/_middleware.dart', on: routesDirectory),
+        exists,
+      );
+    });
+
+    test('Creates middleware in existing dynamic route', () async {
+      await dartFrogNewRoute(
+        '/[existing_dynamic_route]',
+        directory: projectDirectory,
+      );
+
+      expect(
+        fileAt('[existing_dynamic_route].dart', on: routesDirectory),
+        exists,
+      );
+
+      await dartFrogNewMiddleware(
+        '/[existing_dynamic_route]',
+        directory: projectDirectory,
+      );
+
+      expect(
+        fileAt('[existing_dynamic_route].dart', on: routesDirectory),
+        doesNotExist,
+      );
+
+      expect(
+        fileAt('[existing_dynamic_route]/index.dart', on: routesDirectory),
+        exists,
+      );
+
+      expect(
+        fileAt(
+          '[existing_dynamic_route]/_middleware.dart',
+          on: routesDirectory,
+        ),
+        exists,
+      );
+    });
+
+    test('Excuse existing middlewares', () async {
+      await dartFrogNewMiddleware(
+        '/existing_middleware',
+        directory: projectDirectory,
+      );
+
+      await expectLater(
+        () async => dartFrogNewMiddleware(
+          '/existing_middleware',
+          directory: projectDirectory,
+        ),
+        failsWithA(
+          message: 'There is already a middleware at '
+              'routes${slash}existing_middleware${slash}_middleware.dart',
+        ),
+      );
+    });
+
+    test('Excuse middleware creation of invalid route identifier', () async {
+      await expectLater(
+        () async => dartFrogNewMiddleware('/👯‍', directory: projectDirectory),
+        failsWithA(
+          message: 'Route path must be a valid path segment.\n\n',
```

Comment:
I was thinking we can have these message strings as:

```dart
class NewCommand {
...
@visibleForTesting
static const invalidPathSegmentErrorMessage = 'Route path must be a valid path segment.\n\n'
...
}
```

So we avoid duplicating the same string over and over again making it easier to change.

This is a NIT, so feel free to close the comment.

Reply:
Yeah, i like to go DAMP over DRY on tests, but that is not an absolute dogma.